persist/sqlite/hosts.go

```diff
@@ -0,0 +1,315 @@
+					return err
+				}
+				sV4.ProtocolVersion = [3]uint8(protocolVersion)
+
+				if host.V2 {
+					v2AddrRows, err := v2AddrStmt.Query(encode(host.PublicKey))
+					if err != nil {
+						return err
+					}
+					defer v2AddrRows.Close()
+					for v2AddrRows.Next() {
+						var netAddr chain.NetAddress
+						if err := v2AddrRows.Scan(&netAddr.Protocol, &netAddr.Address); err != nil {
+							return err
+						}
+						host.V2NetAddresses = append(host.V2NetAddresses, netAddr)
+					}
+				}
+
+				result = append(result, host)
+				return nil
+			}(); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return
+}
+
+// HostsForScanning returns hosts ordered by the transaction they were created in.
+// Note that only the PublicKey, V2, NetAddress, and V2NetAddresses fields are
+// populated.
+func (s *Store) HostsForScanning(maxLastScan, minLastAnnouncement time.Time, offset, limit uint64) (result []explorer.Host, err error) {
+	err = s.transaction(func(tx *txn) error {
+		rows, err := tx.Query(`SELECT public_key, v2, net_address FROM host_info WHERE last_scan <= ? AND last_announcement >= ? ORDER BY last_scan ASC LIMIT ? OFFSET ?`, encode(maxLastScan), encode(minLastAnnouncement), limit, offset)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+
+		v2AddrStmt, err := tx.Prepare(`SELECT protocol,address FROM host_info_v2_netaddresses WHERE public_key = ? ORDER BY netaddress_order`)
+		if err != nil {
+			return err
+		}
+		defer v2AddrStmt.Close()
+
+		for rows.Next() {
+			var host explorer.Host
+			if err := rows.Scan(decode(&host.PublicKey), &host.V2, &host.NetAddress); err != nil {
+				return err
+			}
+
+			if host.V2 {
+				err := func() error {
+					v2AddrRows, err := v2AddrStmt.Query(encode(host.PublicKey))
+					if err != nil {
+						return err
+					}
+					defer v2AddrRows.Close()
+					for v2AddrRows.Next() {
+						var netAddr chain.NetAddress
+						if err := v2AddrRows.Scan(&netAddr.Protocol, &netAddr.Address); err != nil {
+							return err
+						}
+						host.V2NetAddresses = append(host.V2NetAddresses, netAddr)
+					}
+					return nil
+				}()
+				if err != nil {
+					return err
+				}
+			}
+			result = append(result, host)
+		}
+		return nil
+	})
+	return
+}
+
+// QueryHosts returns the hosts matching the query parameters in the order
+// specified by dir.
+func (st *Store) QueryHosts(params explorer.HostQuery, sortBy explorer.HostSortColumn, dir explorer.HostSortDir, offset, limit uint64) (result []explorer.Host, err error) {
```

Comment:
Why not merge this with the existing `Hosts` endpoint? You already filter by potentially a lot of fields here so you could also filter by public keys and avoid duplicating some of the code.

Reply:
Is there a use case where people want to filter to get a subset of the pubkeys provided versus just a subset of all hosts?  Also, the filtering is v1/v2 different for v1/v2 hosts, so the problem is that someone could have to make two requests to get a host if they are care about hosts with a certain pubkey: one where `params.V2` is false and one where `params.V2` is true.

Reply:
What if we make the V2 param optional, so it can either be unset, true or false. Merging these was the first thing that came to mind when I saw [this](https://github.com/SiaFoundation/explored/pull/152/files#diff-a5457320f6b67f92cbcb51459dd5ebd160912638c6b22d0964e52921b58a05d9R74).

Reply:
Oh I see the problem now. You can solve it but it requires a little rewrite... you essentially have to change all your `if params.V2 then rhp4_setting_x = y else setting_x=y` to `CASE WHEN v2=1 THEN ... ELSE .. END`. That will clean up building the filters and allow you to return both v1 and v2 hosts.